{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "24b1ea39-aaf7-4294-b2eb-e4c3ce635071",
   "metadata": {},
   "source": [
    "This is a pseudocode workflow for reproducing the analysis; for a detailed tutorial, please visit\n",
    "https://starlitnightly.github.io/omicverse/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0765d5c9-e21d-40d6-a545-9dabed3cd474",
   "metadata": {},
   "source": [
    "Bulk RNA-seq process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a1c8087-0928-4f86-98d4-c8853b862146",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Dump\n",
    "fasterq-dump filename -e 32 2> filename_dump.log\n",
    "#QC\n",
    "fastp -i filename.fastq -o filename.fq.gz -f 2 -t 1 -q 15 -u 20 -n 0 -l 35 -w 16 -j filename_fastp.json -h filename_fastp.html -R filename fastp report\n",
    "#Mapping\n",
    "hisat2 -p 55 --dta -x genome_Index -U filename.fq.gz --un-conc-gz filename.fq.gz --new-summary --summary-file filename.align.log | samtools sort -@ 24 -m 6G -o filename.bam\n",
    "#Assembly\n",
    "stringtie filename -o filename.gtf -G genomic.gtf\n",
    "taco_run -v -p 55 --gtf-expr-attr cov --filter-min-expr 2 --filter-splice-juncs -p 55 --ref-genome-fasta genomic_fasta -o TACO GTFs.txt 2> TACO.log\n",
    "#Transcripts filter\n",
    "gffcompare -r genomic.gtf assembly.gtf\n",
    "#Quant\n",
    "align_and_estimate_abundance.pl --transcripts Transcripts.fasta --seqType fq --gene_trans_map Transcripts.fasta.gene_trans_map --samples_file samples.txt --est_method salmon --thread_count 32 --prep_reference\n",
    "abundance_estimates_to_matrix.pl --est_method salmon --gene_trans_map Transcripts.fasta.gene_trans_map --out_prefix Transcripts --name_sample_by_basedir Salmon/*/quant.sf 2> matrix.log\n",
    "#Sequence GO annotation\n",
    "blastp -query Transcripts.fasta.transdecoder_dir/longest_orfs.pep -db  uniprot_sprot.pep -out blastp.outfmt6 -max_target_seqs 1 -outfmt 6 -evalue 1e-5 -num_threads 32\n",
    "hmmsearch --cpu 8 --domtblout pfam.domtblout Pfam-A.hmm Transcripts.fasta.transdecoder_dir/longest_orfs.pep \n",
    "Trinotate-Trinotate-v4.0.2/util/extract_GO_assignments_from_Trinotate_xls.pl --Trinotate_xls Trinotate.xls --G > go_annotations.txt\n",
    "emapper.py -i Transcripts.fasta.transdecoder.pep -o output -m diamond --dmnd_ignore_warnings --cpu 55 --sensmode very-sensitive --id 40 --data_dir dir --dmnd_db eggnog_proteins.dmnd --evalue 1e-5\n",
    "#TF annotation\n",
    "hmmsearch --cpu 8 --domtblout pfam.domtblout TF.hmm Transcripts.fasta.transdecoder_dir/longest_orfs.pep \n",
    "diamond blastx --db /TFDB/TFa.dmnd --query ../NewGTF/Transcripts.fasta -e 1e-5 --outfmt 6 --quiet --very-sensitive --max-target-seqs 1 --id 60 --out TF_predict.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ea0703-7252-4b23-bb15-f5cc2c5d9abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#WGCNA in R\n",
    "library(WGCNA)\n",
    "\n",
    "setwd(\"\")\n",
    "femData<- read.table(\"Transcripts.gene.TPM.not_cross_norm\", header = TRUE, row.names = 1)\n",
    "femData <- femData[!grepl(\"^gene-Trna\", rownames(femData)), ]\n",
    "femData= femData[rowSums(femData>=5) >=2, ] \n",
    "femData <- log2(femData + 1)\n",
    "var_values <- apply(femData, 1, var)\n",
    "cutoff <- quantile(var_values, 0.25)\n",
    "femData <- femData[var_values >= cutoff, ]\n",
    "datExpr0 = as.data.frame(t(femData)) \n",
    "powers = c(c(1:10), seq(from = 12, to=20, by=2))\n",
    "sft = pickSoftThreshold(datExpr0, powerVector = powers, verbose = 5)\n",
    "sizeGrWindow(9, 5)\n",
    "par(mfrow = c(1,2));\n",
    "cex1 = 1;\n",
    "plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],\n",
    "     xlab=\"Soft Threshold (power)\",ylab=\"Scale Free Topology Model Fit,signed R^2\",type=\"n\",\n",
    "     main = paste(\"Scale independence\"));\n",
    "text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],\n",
    "     labels=powers,cex=cex1,col=\"red\");\n",
    "abline(h=0.8,col=\"red\") \n",
    "plot(sft$fitIndices[,1], sft$fitIndices[,5],\n",
    "     xlab=\"Soft Threshold (power)\",ylab=\"Mean Connectivity\", type=\"n\",\n",
    "     main = paste(\"Mean connectivity\"))\n",
    "text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col=\"red\") \n",
    "\n",
    "datExpr0[] <- lapply( datExpr0, as.numeric)\n",
    "WGCNA_matrix<-datExpr0\n",
    "enableWGCNAThreads()\n",
    "#\n",
    "TOM <- TOMsimilarityFromExpr(WGCNA_matrix,power = 8,\n",
    "                             networkType = \"unsigned\",\n",
    "                             nThreads = 50)\n",
    "dissTOM <- 1-TOM\n",
    "geneTree <- hclust(as.dist(dissTOM),method = \"average\")\n",
    "minModuleSize <-100# \n",
    "dynamicMods<- cutreeDynamic(\n",
    "  dendro= geneTree,\n",
    "  distM= dissTOM,\n",
    "  deepSplit= 4,\n",
    "  pamStage= TRUE,\n",
    "  pamRespectsDendro= FALSE,\n",
    "  minClusterSize= minModuleSize)\n",
    "dynamicColors =labels2colors(dynamicMods)\n",
    "p1<-plotDendroAndColors(\n",
    "  geneTree,\n",
    "  dynamicColors,\n",
    "  \"DynamicTree Cut\",\n",
    "  dendroLabels= FALSE, hang = 0.03,\n",
    "  addGuide= TRUE,\n",
    "  guideHang= 0.05,\n",
    "  main= \"Gene dendrogram and module colors\")\n",
    "print(p1)\n",
    "dev.off()\n",
    "\n",
    "MEList<- moduleEigengenes(WGCNA_matrix,colors = dynamicColors)\n",
    "MEs<- MEList$eigengenes\n",
    "MEDiss<- 1-cor(MEs)\n",
    "METree<- hclust(as.dist(MEDiss), method = \"average\")\n",
    "p1<-plot(METree,\n",
    "         main= \"Clustering of module eigengenes\",\n",
    "         xlab= \"\", sub = \"\")\n",
    "abline(h = 0.2, col = \"red\", lty = 2)\n",
    "print(p1)\n",
    "dev.off()\n",
    "\n",
    "MEDissThres<- 0.2\n",
    "merge<- mergeCloseModules(WGCNA_matrix,dynamicColors,cutHeight= MEDissThres,verbose = 3)\n",
    "mergedColors <- merge$colors\n",
    "table(mergedColors)\n",
    "mergedMEs<- merge$newMEs\n",
    "p1<-plotDendroAndColors(\n",
    "  geneTree,\n",
    "  cbind(dynamicColors,mergedColors),\n",
    "  c(\"DynamicTree Cut\", \"Merged dynamic\"),\n",
    "  dendroLabels= FALSE,\n",
    "  hang= 0.03,\n",
    "  addGuide= TRUE,\n",
    "  guideHang= 0.05)\n",
    "print(p1)\n",
    "dev.off()\n",
    "\n",
    "table(mergedColors)\n",
    "write.table(table(mergedColors), file = \"Mergedcolors.txt\", sep = \"\\t\")\n",
    "\n",
    "MEDiss<- 1-cor(mergedMEs)\n",
    "METree <- hclust(as.dist(MEDiss),method = \"average\")\n",
    "par(mar = c(1, 4, 3, 2))\n",
    "p1<-plot(METree,\n",
    "         main= \"Clustering of module eigengenes\",\n",
    "         xlab= \"\", sub = \"\")\n",
    "print(p1)\n",
    "\n",
    "datME=moduleEigengenes(WGCNA_matrix,mergedColors)[[1]]\n",
    "color1=as.character(mergedColors)\n",
    "datKME=signedKME(WGCNA_matrix, datME)\n",
    "dataExpr1=as.data.frame(t(WGCNA_matrix))\n",
    "datSummary=rownames(dataExpr1)\n",
    "datout=data.frame(datSummary,colorNEW=color1,datKME )\n",
    "write.table(datout, \"kMEs.xls\", sep=\"\\t\", row.names=F,quote=F)\n",
    "\n",
    "dir.create('cytoscape')\n",
    "allModules = unique(mergedColors)\n",
    "probes= colnames(WGCNA_matrix)\n",
    "for (module in allModules) {\n",
    "  inModule = is.finite(match(mergedColors, module))\n",
    "  modProbes = probes[inModule]\n",
    "  modTOM = TOM[inModule, inModule]\n",
    "  dimnames(modTOM) = list(modProbes, modProbes)\n",
    "  cyt = exportNetworkToCytoscape(modTOM,\n",
    "                                 edgeFile = paste(\"cytoscape/CytoscapeInput-edges-\", module, \".txt\", sep=\"\"),\n",
    "                                 nodeFile = paste(\"cytoscape/CytoscapeInput-nodes-\", module, \".txt\", sep=\"\"),\n",
    "                                 weighted = TRUE,threshold = 0.02,nodeNames = modProbes,nodeAttr = mergedColors[inModule])\n",
    "}\n",
    "\n",
    "Alldegrees <- intramodularConnectivity(TOM, mergedColors)\n",
    "Alldegrees <- data.frame(\n",
    "  Gene = colnames(WGCNA_matrix), \n",
    "  Module = mergedColors,\n",
    "  Alldegrees\n",
    ")\n",
    "write.table(Alldegrees,\"connetivity.xls\", sep=\"\\t\", row.names=T,quote=F)\n",
    "\n",
    "library(reshape2)\n",
    "samples <- read.table(\"./samples1.txt\", header = FALSE, stringsAsFactors = FALSE)\n",
    "df <- dcast(samples, V2 ~ V1, value.var = \"V1\", fill = 0)\n",
    "rownames(df) <- df$V2\n",
    "df <- df[, -1] \n",
    "df[is.na(df)] <- 0\n",
    "df[ df !=0] <- 1\n",
    "print(df)\n",
    "\n",
    "traitData = df\n",
    "allTraits = traitData\n",
    "\n",
    "traitSamples =rownames(allTraits)\n",
    "expSamples = rownames(datExpr0)\n",
    "traitRows = match(expSamples, traitSamples)\n",
    "datTraits = allTraits[traitRows, ]\n",
    "rownames(datTraits) \n",
    "nSamples = nrow(WGCNA_matrix);\n",
    "moduleTraitCor = cor(mergedMEs, datTraits, use = \"p\");\n",
    "moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a37abf9b-134c-4d25-bca5-40c6c645e00c",
   "metadata": {},
   "source": [
    "ScRNA-seq precess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ad7b9cb-fc5f-4fd2-92d9-6a24c2993d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellranger mkref --genome=genome --fasta=genome_asta --genes=genomic.gtf\n",
    "cellranger count --id=genome --transcriptome=reference --fastqs=fastq_files --sample=sample --localcores=24 --localmem=256"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "511da89d-ccbd-48c1-ad3f-bea82a9ad739",
   "metadata": {},
   "outputs": [],
   "source": [
    "import omicverse as ov\n",
    "print(f\"omiverse version: {ov.__version__}\")\n",
    "import scanpy as sc\n",
    "print(f\"scanpy version: {sc.__version__}\")\n",
    "ov.ov_plot_set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b11d937-0e2e-40c5-9c3b-cee46261bf71",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Merge data\n",
    "adata11 = sc.read_10x_h5(filename=\"blood/SRR18899520/wssvblood/outs/filtered_feature_bc_matrix.h5\")\n",
    "adata11.obs['tissue']='Blood' \n",
    "adata11.obs['project']='p1' \n",
    "adata11.obs['batch']='batch1' \n",
    "adata11.obs['condiction']='wssv'\n",
    "adata1 = sc.read_10x_h5(filename=\"blood/SRR18899521/bloodout/outs/filtered_feature_bc_matrix.h5\")\n",
    "adata1.obs['tissue']='Blood' \n",
    "adata1.obs['project']='p1' \n",
    "adata1.obs['batch']='batch1' \n",
    "adata1.obs['condiction']='normal'\n",
    "...\n",
    "adata=sc.concat([adata1,...adata6,adata7,adata8,adata9,adata10,adata11],join='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86d1b0b4-30bd-476a-b6f1-e9f3a48be7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#QC\n",
    "adata.var_names_make_unique()\n",
    "adata.obs_names_make_unique()\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"KEG\")\n",
    "from scipy.sparse import issparse\n",
    "if issparse(adata.X):\n",
    "    adata.obs['nUMIs'] = adata.X.toarray().sum(axis=1)\n",
    "    adata.obs['mito_perc'] = adata[:, adata.var[\"mt\"]].X.toarray().sum(axis=1) / adata.obs['nUMIs'].values\n",
    "    adata.obs['detected_genes'] = (adata.X.toarray() > 0).sum(axis=1)\n",
    "else:\n",
    "    adata.obs['nUMIs'] = adata.X.sum(axis=1)\n",
    "    adata.obs['mito_perc'] = adata[:, adata.var[\"mt\"]].X.sum(axis=1) / adata.obs['nUMIs'].values\n",
    "    adata.obs['detected_genes'] = (adata.X > 0).sum(axis=1)\n",
    "adata = adata[adata.obs['mito_perc'] <= 0.2]\n",
    "qc_thresholds = {'mito_perc': 0.2, 'nUMIs': 500, 'detected_genes': 250}\n",
    "adata = ov.pp.qc(adata, tresh=qc_thresholds,batch_key='batch')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0dc2174-a1ee-47f1-9321-e707ef31a50d",
   "metadata": {},
   "source": [
    "Annotation with known marker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9df992-6c19-4660-89a1-120ab4aca984",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Hemocytes annotation\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "fig, axes = plt.subplots(3, 3, figsize=(18, 18))\n",
    "cmap = 'Reds'\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113823934', show=False, ax=axes[0, 0], cmap=cmap)#TGase1 \n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113823945', show=False, ax=axes[0, 1], cmap=cmap)#TGase\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113814237', show=False, ax=axes[0, 2], cmap=cmap)#Notch1\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113804553', show=False, ax=axes[1, 0], cmap=cmap)#CTL2 \n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113809352', show=False, ax=axes[1, 1], cmap=cmap)#SVC5\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113822052', show=False, ax=axes[1, 2], cmap=cmap)#Ast2\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113802754', show=False, ax=axes[2, 0], cmap=cmap)#CHF\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113816267', show=False, ax=axes[2, 1], cmap=cmap)#CrustinL\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113801825', show=False, ax=axes[2, 2], cmap=cmap)#Crustin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c16e6152-61ff-4546-b372-97aac6d11350",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Macrophage-like hemocytes subpopulation \n",
    "'''\n",
    "CHIT1 LOC113822241\n",
    "Lyz1 LOC113802295\n",
    "LIPF LOC113823443\n",
    "LGMN LOC113827868\n",
    "Nlrp3 LOC113828560\n",
    "NAGA LOC113810366\n",
    "zfh1 LOC113824903\n",
    "Casp1 LOC113800234\n",
    "NPC2 LOC113806913\n",
    "'''\n",
    "fig, axes = plt.subplots(3, 3, figsize=(18, 18))\n",
    "cmap = 'Purples'\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113822241', show=False, ax=axes[0, 0], cmap=cmap) \n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113802295', show=False, ax=axes[0, 1], cmap=cmap)\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113823443', show=False, ax=axes[0, 2], cmap=cmap)\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113827868', show=False, ax=axes[1, 0], cmap=cmap)\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113828560', show=False, ax=axes[1, 1], cmap=cmap)\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113810366', show=False, ax=axes[1, 2], cmap=cmap)\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113824903', show=False, ax=axes[2, 0], cmap=cmap)\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113800234', show=False, ax=axes[2, 1], cmap=cmap)\n",
    "ov.utils.embedding(adat2, basis='X_umap', color='LOC113806913', show=False, ax=axes[2, 2], cmap=cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8180cf89-e13b-481b-8e03-7970e0ff4b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Hep\n",
    "#E cell\n",
    "fig, axes = plt.subplots(2, 3, figsize=(16, 10))\n",
    "cmap = 'Blues'\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113819379', show=False, ax=axes[0, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113823098', show=False, ax=axes[0, 1], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113819117', show=False, ax=axes[0, 2], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113823089', show=False, ax=axes[1, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113827499', show=False, ax=axes[1, 1], cmap=cmap)\n",
    "#B cell\n",
    "fig, axes = plt.subplots(2, 3, figsize=(16, 10))\n",
    "cmap = 'Blues'\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113823145', show=False, ax=axes[0, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113809598', show=False, ax=axes[0, 1], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113822815', show=False, ax=axes[0, 2], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113810775', show=False, ax=axes[1, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113804689', show=False, ax=axes[1, 1], cmap=cmap)\n",
    "#F cell \n",
    "fig, axes = plt.subplots(2, 3, figsize=(16, 10))\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113819352', show=False, ax=axes[0, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113817198', show=False, ax=axes[0, 1], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113811244', show=False, ax=axes[0, 2], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113814344', show=False, ax=axes[1, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113819139', show=False, ax=axes[1, 1], cmap=cmap)\n",
    "#F cell \n",
    "fig, axes = plt.subplots(2, 3, figsize=(16, 10))\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113809896', show=False, ax=axes[0, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113820201', show=False, ax=axes[0, 1], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113819222', show=False, ax=axes[0, 2], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113806454', show=False, ax=axes[1, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113813404', show=False, ax=axes[1, 1], cmap=cmap)\n",
    "#R cell\n",
    "fig, axes = plt.subplots(2, 3, figsize=(16, 10))\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113817261', show=False, ax=axes[0, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113810053', show=False, ax=axes[0, 1], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113799989', show=False, ax=axes[0, 2], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113817260', show=False, ax=axes[1, 0], cmap=cmap)\n",
    "ov.utils.embedding(adata, basis='X_umap', color='LOC113810057', show=False, ax=axes[1, 1], cmap=cmap)\n",
    "genes_of_interest = [#E\n",
    " \n",
    "'LOC113819379',\n",
    "'LOC113823098',\n",
    "'LOC113819117',\n",
    "'LOC113823089',\n",
    "'LOC113827499',\n",
    "    #R\n",
    "    'LOC113817261', \n",
    "'LOC113810053',\n",
    "'LOC113799989',\n",
    "'LOC113817260',\n",
    "'LOC113810057', \n",
    "    #F\n",
    "  'LOC113809896',\n",
    "'LOC113820201',\n",
    "'LOC113819222',\n",
    "'LOC113806454',\n",
    "'LOC113813404',\n",
    "'LOC113819352',\n",
    "'LOC113811244',\n",
    "'LOC113814344',\n",
    "'LOC113819139',\n",
    "  #B\n",
    "'LOC113823145',\n",
    "'LOC113809598',\n",
    "'LOC113822815',\n",
    "'LOC113810775',\n",
    "'LOC113804689',\n",
    "#CTL\n",
    "    'LOC113804553',\n",
    "'LOC113809352',\n",
    "'LOC113816267', \n",
    "'LOC113801825', \n",
    "]\n",
    "sc.pl.dotplot(\n",
    "    adata,\n",
    "    var_names=genes_of_interest, \n",
    "    groupby='leiden_res', \n",
    "    cmap='Blues',\n",
    "    standard_scale='var'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dc7e318-8b26-4a23-839c-f1e8c05439d7",
   "metadata": {},
   "source": [
    "Novel cell type annotaion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b1cacc8-3e67-4d5b-9250-f3e1d997b581",
   "metadata": {},
   "outputs": [],
   "source": [
    "##DEG\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata, groupby=\"major_celltype\", use_raw=False,\n",
    "    method=\"t-test\",)\n",
    "groups = adata.uns['rank_genes_groups']['names'].dtype.names\n",
    "merged_df = pd.DataFrame()\n",
    "for group in groups:\n",
    "    ranked_genes_df = sc.get.rank_genes_groups_df(adata, group=group)\n",
    "    ranked_genes_df = ranked_genes_df[['names', 'scores']]\n",
    "    ranked_genes_df.rename(columns={'scores': f'{group}_score'}, inplace=True)\n",
    "    if merged_df.empty:\n",
    "        merged_df = ranked_genes_df\n",
    "    else:\n",
    "        merged_df = pd.merge(merged_df, ranked_genes_df, on='names', how='outer')\n",
    "merged_df.to_csv ('ranksocre.txt',sep='\\t')\n",
    "groups = adata.uns['rank_genes_groups']['names'].dtype.names\n",
    "merged_df = pd.DataFrame()\n",
    "for group in groups:\n",
    "    ranked_genes_df = sc.get.rank_genes_groups_df(adata, group=group)\n",
    "    ranked_genes_df = ranked_genes_df[['names', 'logfoldchanges']]\n",
    "    ranked_genes_df.rename(columns={'logfoldchanges': f'{group}_lgFC'}, inplace=True)\n",
    "    if merged_df.empty:\n",
    "        merged_df = ranked_genes_df\n",
    "    else:\n",
    "        merged_df = pd.merge(merged_df, ranked_genes_df, on='names', how='outer')\n",
    "merged_df.to_csv ('rankFC.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "015d4d35-7edc-41e0-8bcf-ae8a3afe818d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#AUCell for GO\n",
    "pathway_dict=ov.utils.geneset_prepare('output.gmt')\n",
    "adata_aucs=ov.single.pathway_aucell_enrichment(adata,pathways_dict=pathway_dict, num_workers=48)\n",
    "adata_aucs.obs=adata[adata_aucs.obs.index].obs\n",
    "adata_aucs.obsm=adata[adata_aucs.obs.index].obsm\n",
    "adata_aucs.obsp=adata[adata_aucs.obs.index].obsp\n",
    "sc.tl.dendrogram(adata_aucs,'major_celltype',use_rep=\"X_umap\")\n",
    "sc.tl.rank_genes_groups(adata_aucs, 'major_celltype', method='t-test',n_genes=100)\n",
    "sc.tl.rank_genes_groups(adata_aucs, 'major_celltype', method='t-test')\n",
    "import pandas as pd\n",
    "adata=adata_aucs\n",
    "groups = adata.uns['rank_genes_groups']['names'].dtype.names\n",
    "merged_df = pd.DataFrame()\n",
    "for group in groups:\n",
    "    ranked_genes_df = sc.get.rank_genes_groups_df(adata, group=group)\n",
    "    ranked_genes_df = ranked_genes_df[['names', 'scores']]\n",
    "    ranked_genes_df.rename(columns={'scores': f'{group}_score'}, inplace=True)\n",
    "    if merged_df.empty:\n",
    "        merged_df = ranked_genes_df\n",
    "    else:\n",
    "        merged_df = pd.merge(merged_df, ranked_genes_df, on='names', how='outer')\n",
    "merged_df.to_csv ('subranksocre.txt',sep='\\t')\n",
    "groups = adata.uns['rank_genes_groups']['names'].dtype.names\n",
    "merged_df = pd.DataFrame()\n",
    "for group in groups:\n",
    "    ranked_genes_df = sc.get.rank_genes_groups_df(adata, group=group)\n",
    "    ranked_genes_df = ranked_genes_df[['names', 'logfoldchanges']]\n",
    "    ranked_genes_df.rename(columns={'logfoldchanges': f'{group}_lgFC'}, inplace=True)\n",
    "    if merged_df.empty:\n",
    "        merged_df = ranked_genes_df\n",
    "    else:\n",
    "        merged_df = pd.merge(merged_df, ranked_genes_df, on='names', how='outer')\n",
    "merged_df.to_csv ('subrankFC.txt',sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (omicverse)",
   "language": "python",
   "name": "omicverse"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
